nomad/job_endpoint_validators.go

```diff
@@ -0,0 +1,56 @@
+package nomad
+
+import (
+	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/pkg/errors"
+)
+
+type jobNamespaceConstraintCheckHook struct {
+	srv *Server
+}
+
+func (jobNamespaceConstraintCheckHook) Name() string {
+	return "namespace-constraint-check"
+}
+
+func (c jobNamespaceConstraintCheckHook) Validate(job *structs.Job) (warnings []error, err error) {
+	// This was validated before and matches the WriteRequest namespace
+	ns, err := c.srv.State().NamespaceByName(nil, job.Namespace)
+	if err != nil {
+		return nil, err
+	}
+	if ns == nil {
+		return nil, errors.Errorf("requested namespace %s does not exist", job.Namespace)
```

Comment:
> One way to fix it would be to adjust my error message about non-existing namespace to match the one `upsertJobImpl` raises but then I am not sure if the existing test checks the wanted state…

Arguably the test is checking the correct state but at the wrong spot, and we should have it in validation. But if we _only_ checked it during validation the namespace could be deleted out from under us while processing the RPC but before the transaction was opened for the upsert. The state store has its own tests so from the perspective of the job endpoint test it shouldn't matter where the error is bubbling up from.

The language of the error messages isn't consistent across the endpoints unfortunately. For this PR let's just use the same language as the state store error so that we can unblock this feature.

Reply:
I have fixed that and would like to add a test for the validate method. Can you recommend where (I also need to configure some state since it accesses the namespace… mainly looking for examples where this is already done so I can copy & paste :D)

Reply:
A new file named `job_endpoint_validators_test.go` would be the best place for it. [`TestNamespaceEndpoint_GetNamespace`](https://github.com/hashicorp/nomad/blob/main/nomad/namespace_endpoint_test.go#L17) has a reasonable example of setting up test state.